### PR TITLE
Add per-player time offset

### PR DESCRIPTION
**Pull request by Arcelmi**
With this pull request it's possible to change the timeOfDay of a certain player for e.g. 1 hour. So it's possible to make in-game time zones or something like this.
Unlike override_day_night_ratio it also change the position of the moon and sun and not only the sunlight. And it isn't static like override_day_night_ratio and is relative to the timeOfDay.
It's also possible to make machines that change the time in a certain radius and many other possibilities. It's also useful on minigame-servers when players in different minigames have got different daytime.

### CODE REVIEW

**Comment by sfan5** [comment]
interesting :thinking:

File: doc/lua_api.txt [hunk 1]
```diff
@@ -3783,6 +3783,9 @@ This is basically a reference to a C++ `ServerActiveObject`
     * `0`...`1`: Overrides day-night ratio, controlling sunlight to a specific amount
     * `nil`: Disables override, defaulting to sunlight based on day-night cycle
 * `get_day_night_ratio()`: returns the ratio or nil if it isn't overridden
+* `set_time_offset(offset)`
+	* Adds offset to the time of the player (e.g. TimeOfDay is 12:00 and the offset of the player is 1000 then the time of the player is 13:00)
```
Comment (sfan5) [inline doc/lua_api.txt]: put the example on a separate line

File: doc/lua_api.txt [hunk 2]
```diff
@@ -3783,6 +3783,9 @@ This is basically a reference to a C++ `ServerActiveObject`
...
+* `get_time_offset()` returns the offset
```
Comment (sfan5) [inline doc/lua_api.txt]: <pre>* `get_time_offset()`: returns the players time of day offset</pre>
would be better

File: src/remoteplayer.h [hunk 3]
```diff
@@ -69,6 +69,16 @@ class RemotePlayer : public Player
 		*ratio = m_day_night_ratio;
 	}
 
+	void setTimeOffset(u16 offset)
+	{
+		m_time_offset = offset;
+	}
+
+	void getTimeOffset(u16 *offset)
```
Comment (sfan5) [inline src/remoteplayer.h]: is there a reason why this isn't `u16 getTimeOffset() const`?

File: src/script/lua_api/l_object.cpp [hunk 4]
```diff
@@ -1720,6 +1720,45 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
...
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	RemotePlayer *player = getplayer(ref);
+	if (player == NULL)
+		return 0;
+
+	int offset_i = luaL_checknumber(L, 2);
+	u16 offset = offset_i;
+	//Negative values
```
Comment (sfan5) [inline src/script/lua_api/l_object.cpp]: why is the offset not just a signed number (`s16`)?

File: src/script/lua_api/l_object.cpp [hunk 5]
```diff
@@ -1720,6 +1720,45 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
...
+	if (offset_i<0){
+		offset = 24000+offset_i;
+	}
+
+	player->setTimeOffset(offset);
+	getServer(L)->setTimeOfDay(getEnv(L)->getTimeOfDay());
```
Comment (sfan5) [inline src/script/lua_api/l_object.cpp]: better just resend the time to that one player

File: src/server.cpp [hunk 6]
```diff
@@ -1800,13 +1800,22 @@ void Server::SendOverrideDayNightRatio(session_t peer_id, bool do_override,
...
+			u16 time_offset;
+			player->getTimeOffset(&time_offset);
+			u16 ntime = (time+time_offset);
+			pkt << ntime << time_speed;
+			Send(&pkt);
+		}
 	}
 	else {
+		NetworkPacket pkt(TOCLIENT_TIME_OF_DAY, 0, peer_id);
+		pkt << time << time_speed;
```
Comment (sfan5) [inline src/server.cpp]: this code does not account for time offsets

File: src/server.cpp [hunk 7]
```diff
@@ -1800,13 +1800,22 @@ void Server::SendOverrideDayNightRatio(session_t peer_id, bool do_override,
...
-		m_clients.sendToAll(&pkt);
+		for (auto &client_name : m_clients.getPlayerNames()) {
+			RemotePlayer *player = m_env->getPlayer(client_name.c_str());
+			peer_id = player->getPeerId();
+			NetworkPacket pkt(TOCLIENT_TIME_OF_DAY, 0, peer_id);
+			u16 time_offset;
+			player->getTimeOffset(&time_offset);
+			u16 ntime = (time+time_offset);
+			pkt << ntime << time_speed;
+			Send(&pkt);
```
Comment (sfan5) [inline src/server.cpp]: you could move the actual implementation into the `else` and have this code only call `SendTimeOfDay(peer_id, time, time_speed)` for each player

File: src/remoteplayer.h (hunk 3 above)
Comment (red-001) [reply src/remoteplayer.h]: Why not just return the number instead of using a pointer?

Comment (Arcelmi) [reply src/remoteplayer.h]: I used getDayNightRatio as an example because I'm not so good in c++.

File: src/script/lua_api/l_object.cpp (hunk 4 above)
Comment (Arcelmi) [reply src/script/lua_api/l_object.cpp]: Because timeofday is an unsigned number and when I add a negative offset to timeofday and timeofday<0 then it would be the wrong time.

**Comment by Ekdohibs** [comment]
While you're doing this, could you add per-player time speed as well please?

File: src/server.cpp [hunk 8]
```diff
@@ -1800,13 +1800,28 @@ void Server::SendOverrideDayNightRatio(session_t peer_id, bool do_override,
...
 {
-	NetworkPacket pkt(TOCLIENT_TIME_OF_DAY, 0, peer_id);
-	pkt << time << time_speed;
 
 	if (peer_id == PEER_ID_INEXISTENT) {
-		m_clients.sendToAll(&pkt);
+		for (auto &client_name : m_clients.getPlayerNames()) {
+			RemotePlayer *player = m_env->getPlayer(client_name.c_str());
+			peer_id = player->getPeerId();
+			NetworkPacket pkt(TOCLIENT_TIME_OF_DAY, 0, peer_id);
```
Comment (rubenwardy) [inline src/server.cpp]: please add a blank line above here

File: doc/lua_api.txt [hunk 9]
```diff
@@ -3783,6 +3783,10 @@ This is basically a reference to a C++ `ServerActiveObject`
     * `0`...`1`: Overrides day-night ratio, controlling sunlight to a specific amount
     * `nil`: Disables override, defaulting to sunlight based on day-night cycle
 * `get_day_night_ratio()`: returns the ratio or nil if it isn't overridden
+* `set_time_offset(offset)`
+	* Adds offset to the time of the player 
+	* e.g. TimeOfDay is 12:00 and the offset of the player is 1000 then the time of the player is 13:00
```
Comment (SmallJoker) [inline doc/lua_api.txt]: wrong indents

File: src/script/lua_api/l_object.cpp [hunk 10]
```diff
@@ -1720,6 +1720,44 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
...
+	RemotePlayer *player = getplayer(ref);
+	if (player == NULL)
+		return 0;
+
+	int offset_i = luaL_checknumber(L, 2);
+	u16 offset = offset_i;
+	//Negative values
+	if (offset_i<0){
+		offset = 24000+offset_i;
+	}
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp]: Remove brackets or add a space after `)` -> `0) {`

File: src/script/lua_api/l_object.cpp [hunk 11]
```diff
@@ -1720,6 +1720,44 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
...
+	ObjectRef *ref = checkobject(L, 1);
+	RemotePlayer *player = getplayer(ref);
+	if (player == NULL)
+		return 0;
+
+	int offset_i = luaL_checknumber(L, 2);
+	u16 offset = offset_i;
+	//Negative values
+	if (offset_i<0){
+		offset = 24000+offset_i;
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp]: Space between operator
http://dev.minetest.net/Code_style_guidelines

File: src/script/lua_api/l_object.cpp [hunk 12]
```diff
@@ -1720,6 +1720,44 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
...
+	}
+
+	player->setTimeOffset(offset);
+	getServer(L)->SendTimeOfDay(player->getPeerId(), getEnv(L)->getTimeOfDay(), g_settings->getFloat("time_speed"));
+
+	lua_pushboolean(L, true);
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp]: Useless return value as the player object must logically be valid and there's no way to fail otherwise.

File: src/server.cpp [hunk 13]
```diff
@@ -1800,13 +1800,29 @@ void Server::SendOverrideDayNightRatio(session_t peer_id, bool do_override,
...
+			pkt << ntime << time_speed;
+			Send(&pkt);
+		}
 	}
 	else {
+		RemotePlayer *player = m_env->getPlayer(peer_id);
+		// When the player is joining, player is NULL
+		if (!player){
+			return;
+		}
```
Comment (SmallJoker) [inline src/server.cpp]: Same as above. Either remove brackets or put a space after `)`

File: src/script/lua_api/l_object.cpp (hunk 12 above)
Comment (Arcelmi) [reply src/script/lua_api/l_object.cpp]: It returns nil if (player == NULL).

File: doc/lua_api.txt [hunk 14]
```diff
@@ -3783,6 +3783,10 @@ This is basically a reference to a C++ `ServerActiveObject`
     * `0`...`1`: Overrides day-night ratio, controlling sunlight to a specific amount
     * `nil`: Disables override, defaulting to sunlight based on day-night cycle
 * `get_day_night_ratio()`: returns the ratio or nil if it isn't overridden
+* `set_time_offset(offset)`
+	* Adds offset to the time of the player 
+	e.g. TimeOfDay is 12:00 and the offset of the player is 1000 then the time of the player is 13:00
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Apparently because we love inconsistency with the code: the docs are indented with s p a c e s.

File: src/script/lua_api/l_object.cpp [hunk 15]
```diff
@@ -1720,6 +1720,45 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
 	return 1;
 }
 
+// set_time_offset(self, offset)
+int ObjectRef::l_set_time_offset(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	RemotePlayer *player = getplayer(ref);
+	if (player == NULL)
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp]: `!player`

File: src/script/lua_api/l_object.cpp [hunk 16]
```diff
@@ -1720,6 +1720,45 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
...
+		return 0;
+
+	int offset_i = luaL_checknumber(L, 2);
+	u16 offset = offset_i;
+	//Negative values
+	if (offset_i<0) {
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp]: Spaces around operator please.

File: src/script/lua_api/l_object.cpp [hunk 17]
```diff
@@ -1720,6 +1720,45 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
...
+	return 1;
+}
+
+// get_time_offset(self)
+int ObjectRef::l_get_time_offset(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	RemotePlayer *player = getplayer(ref);
+	if (player == NULL)
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp]: `!player`

File: src/server.cpp [hunk 18]
```diff
@@ -1800,13 +1800,31 @@ void Server::SendOverrideDayNightRatio(session_t peer_id, bool do_override,
...
-	NetworkPacket pkt(TOCLIENT_TIME_OF_DAY, 0, peer_id);
-	pkt << time << time_speed;
 
 	if (peer_id == PEER_ID_INEXISTENT) {
-		m_clients.sendToAll(&pkt);
+		std::vector<session_t> clients = m_clients.getClientIDs();
+		for (const session_t peer_id : clients) {
+			RemotePlayer *player = m_env->getPlayer(peer_id);
+			if (!player) {
+				return;
```
Comment (SmallJoker) [inline src/server.cpp]: Don't return. Continue, as the loop shall not be stopped.

File: src/script/lua_api/l_object.cpp (hunk 15 above)
Comment (Arcelmi) [reply src/script/lua_api/l_object.cpp]: Every other function in this file is with player == NULL too.

File: src/script/lua_api/l_object.cpp (hunk 17 above)
Comment (Arcelmi) [reply src/script/lua_api/l_object.cpp]: Every other function in this file is with player == NULL too.

File: src/script/lua_api/l_object.cpp (hunk 15 above)
Comment (SmallJoker) [reply src/script/lua_api/l_object.cpp]: The fact that other functions do the same doesn't mean it's good code.

Comment (Arcelmi) [reply src/script/lua_api/l_object.cpp]: That was my first real C++-Code so I was inspired by the other code.

**Comment by SmallJoker** [comment]
@Arcelmi Could you please satisfy LINT's needs so this can be merged after the 2nd approval? https://travis-ci.org/minetest/minetest/jobs/352884874

Testing mod:
```Lua
minetest.register_chatcommand("timediff", {
	func = function(name, param)
		local player = minetest.get_player_by_name(name)
		player:set_time_offset(tonumber(param) or 0);
	end
end)
```

**Comment by paramat** [comment]
LINT's requests are unacceptable here.
I disagree we should change our codestyle to satisfy LINT, you can add this file to the LINT whitelist instead.

File: doc/lua_api.txt [hunk 19]
```diff
@@ -3783,6 +3783,10 @@ This is basically a reference to a C++ `ServerActiveObject`
     * `0`...`1`: Overrides day-night ratio, controlling sunlight to a specific amount
     * `nil`: Disables override, defaulting to sunlight based on day-night cycle
 * `get_day_night_ratio()`: returns the ratio or nil if it isn't overridden
+* `set_time_offset(offset)`
+    * Adds offset to the time of the player 
+    e.g. TimeOfDay is 12:00 and the offset of the player is 1000 then the time of the player is 13:00
```
Comment (paramat) [inline doc/lua_api.txt]: Best write 12000 and 13000.

**Comment by SmallJoker** [comment]
@paramat So in that case.. merge this PR without regard of LINT as soon it has two approvals? For me that would be fine too.

**Comment by paramat** [comment]
Ok, but best it is added to the whitelist by the author.

**Comment by nerzhul** [comment]
just a note: this totally break mechanics based on time, with that player will have an offset for time based events

File: src/script/lua_api/l_object.cpp (hunk 4 above)
Comment (rdococ) [reply src/script/lua_api/l_object.cpp]: either way, an offset of, say, 23999 would effectively work the same, and if anything you could perform a modulus (t % 24000)

File: src/script/lua_api/l_object.cpp [hunk 20]
```diff
@@ -1720,6 +1720,45 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
...
+	RemotePlayer *player = getplayer(ref);
+	if (!player)
+		return 0;
+
+	int offset_i = luaL_checknumber(L, 2);
+	u16 offset = offset_i;
+	//Negative values
+	if (offset_i < 0) {
+		offset = 24000 + offset_i;
+	}
```
Comment (sfan5) [inline src/script/lua_api/l_object.cpp]: brace removal, also the comment should explain why it's unsigned

File: src/script/lua_api/l_object.cpp [hunk 21]
```diff
@@ -1720,6 +1720,45 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
...
+
+	player->setTimeOffset(offset);
+	getServer(L)->SendTimeOfDay(player->getPeerId(),
```
Comment (sfan5) [inline src/script/lua_api/l_object.cpp]: IMO this should not be done directly from here
instead of making `SendTimeOfDay` public, add a method in Serve such as `updatePlayerTime()` that does this

File: src/server.cpp [hunk 22]
```diff
@@ -1800,13 +1800,30 @@ void Server::SendOverrideDayNightRatio(session_t peer_id, bool do_override,
...
 {
-	NetworkPacket pkt(TOCLIENT_TIME_OF_DAY, 0, peer_id);
-	pkt << time << time_speed;
 
 	if (peer_id == PEER_ID_INEXISTENT) {
-		m_clients.sendToAll(&pkt);
+		std::vector<session_t> clients = m_clients.getClientIDs();
+		for (const session_t peer_id : clients) {
+			RemotePlayer *player = m_env->getPlayer(peer_id);
+			if (player) {
```
Comment (sfan5) [inline src/server.cpp]:
```c
if (!player)
    continue;
[...]
```

File: src/server.cpp [hunk 23]
```diff
@@ -1800,13 +1800,30 @@ void Server::SendOverrideDayNightRatio(session_t peer_id, bool do_override,
...
+				Send(&pkt);
+			}
+		}
 	}
 	else {
+		RemotePlayer *player = m_env->getPlayer(peer_id);
+		// When the player is joining, player is NULL
+		if (!player) {
+			return;
+		}
```
Comment (sfan5) [inline src/server.cpp]: brace removal

File: util/travis/clang-format-whitelist.txt [hunk 24]
```diff
@@ -406,3 +406,5 @@ src/voxelalgorithms.h
 src/voxel.cpp
 src/voxel.h
 src/wieldmesh.cpp
+src/remoteplayer.h
+src/gui/guiConfirmRegistration.cpp
```
Comment (sfan5) [inline util/travis/clang-format-whitelist.txt]: this seems unrelated to this pr, why is this here?

File: doc/lua_api.txt (hunk 19 above)
Comment (paramat) [inline doc/lua_api.txt]: It's clearer to use 12000, 1000, 13000 here. This form of time seems more familiar to modders.

File: util/travis/clang-format-whitelist.txt (hunk 24 above)
Comment (Arcelmi) [reply util/travis/clang-format-whitelist.txt]: Because there was this line in Travis CI build:
"The file src/gui/guiConfirmRegistration.cpp is not compliant with the coding style:" - https://travis-ci.org/minetest/minetest/jobs/352884874
